mfe.go

```diff
@@ -0,0 +1,1092 @@
+	energyContributions := make([]EnergyContribution, 0)
+	energy, contribution := externalLoopEnergy(fc)
+	energyContributions = append(energyContributions, contribution)
+
+	pairTable := fc.pairTable
+	for i := 0; i < fc.length; i++ {
+		if pairTable[i] == -1 {
+			continue
+		}
+
+		energy += stackEnergy(fc, i, &energyContributions)
+		i = pairTable[i]
+	}
+
+	return energy, energyContributions
+}
+
+/*
+*	returns the numerical representation of a base pair based on the map
+* `basePairEncodedTypeMap`. 7 is a special value that signifies a non-standard
+* base pair (see `energy_params.md` for more information).
+* As stated in `basePairEncodedTypeMap()`, the numerical representation carries
+* no meaning in itself except for where to find the relevent energy
+* contributions of the base pair (i, j) in the matrices of the energy
+* paramaters (found in `energy_params.go`).
+ */
+func encodedBasePairType(i, j byte, basePairEncodedTypeMap map[byte]map[byte]int) int {
+	var encodedType int = basePairEncodedTypeMap[i][j]
+
+	if encodedType == 0 {
+		return 7
+	}
+
+	return encodedType
+}
+
+/**
+* Calculate the energy contribution of stabilizing dangling-ends/mismatches
+* for all stems branching off the exterior loop
+ */
+func externalLoopEnergy(fc *foldCompound) (int, EnergyContribution) {
+	pairTable := fc.pairTable
+	var energy int = 0
+	length := fc.length
+	pairFivePrimeIdx := 0
+
+	// seek to opening base of first stem
+	for pairFivePrimeIdx < length && pairTable[pairFivePrimeIdx] == -1 {
+		pairFivePrimeIdx++
+	}
+
+	for pairFivePrimeIdx < length {
+		/* pairFivePrimeIdx must have a pairing partner */
+		pairThreePrimeIdx := pairTable[pairFivePrimeIdx]
+
+		/* get encoded type of base pair (pairFivePrimeIdx, pairThreePrimeIdx) */
+		basePairType := encodedBasePairType(fc.sequence[pairFivePrimeIdx],
+			fc.sequence[pairThreePrimeIdx],
+			fc.energyParams.basePairEncodedTypeMap)
+
+		var fivePrimeMismatch, threePrimeMismatch int
+		if pairFivePrimeIdx > 0 {
+			fivePrimeMismatch = fc.encodedSequence[pairFivePrimeIdx-1]
+		} else {
+			fivePrimeMismatch = -1
+		}
+
+		if pairThreePrimeIdx < length-1 {
+			threePrimeMismatch = fc.encodedSequence[pairThreePrimeIdx+1]
+		} else {
+			threePrimeMismatch = -1
+		}
+
+		energy += exteriorStemEnergy(basePairType, fivePrimeMismatch, threePrimeMismatch, fc.energyParams)
+
+		// seek to the next stem
+		pairFivePrimeIdx = pairThreePrimeIdx + 1
+		for pairFivePrimeIdx < length && pairTable[pairFivePrimeIdx] == -1 {
+			pairFivePrimeIdx++
+		}
+	}
+
+	return energy, EnergyContribution{energy: energy}
+}
+
+/**
+ *  Evaluate a stem branching off the exterior loop.
+ *
+ *  Given a base pair (i,j) (encoded by `basePairType`), compute the
+ *  energy contribution including dangling-end/terminal-mismatch contributions.
+ *  Instead of returning the energy contribution per-se, this function returns
+ *  the corresponding Boltzmann factor.
```

Comment:
What is a Boltzmann factor? Can we get a small review of that here to understand what it is returning?

Reply:
I tried to read up about the Boltzmann factor in their repo and it looks like it has nothing to do with this function. The comment relates to another function, but not sure why it's included here. I've removed it now.